tests/semantic: Add architecture specific testcase support

lostjeffle opened the pull request:
commit 48d405133160 ("bpftrace: Add architecture specific testcase
support") added architecture specific support for runtime tests.
Still there are architecture specific testcases in semantic_analyser,
where 'ip' register is hard coded in semantic_analyser.call_reg and
semantic_analyser.builtin_functions.

These testcases are all c/c++ files and thus we need conditional compiling.

Signed-off-by: Jeffle Xu <jefflexu@linux.alibaba.com>

#1137 had ever noted that some architecture specific test cases are hard coded, and thus will fail in some architectures not supported.

- some runtime-test had been fixed by  commit 48d405133160 ("bpftrace: Add architecture specific testcase support")
- opensnoop/statsnoop in tools-parsing-test had been fixed by commit 330953c11190c23a12f98d374d1b5eb6d7aad1fe ("fix build failed on aarch64")

This patch is supposed to fix the failure in semantic_analyser.builtin_functions, semantic_analyser.call_reg.


##### Checklist

- [ ] Language changes are updated in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

fbs commented on CHANGELOG.md:
```diff
@@ -106,6 +106,8 @@ and this project adheres to
   - [#1580](https://github.com/iovisor/bpftrace/pull/1580)
 - Printing of small integers with `printf`
   - [#1532](https://github.com/iovisor/bpftrace/pull/1532)
+- Add architecture specific testcase support for semantic_analyser
```
We only put the enduser facing changes in the changelog, internal test changes are not relevant for them so you can remove this :)

lostjeffle replied:
Regards.

danobi commented on tests/CMakeLists.txt:
```diff
@@ -2,6 +2,19 @@ add_compile_options("-Wno-undef")
 
+# Architecture-specific preprocessor macro for *.cpp/*.c
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "aarch64")
+  add_definitions(-DARCH_AARCH64)
```
I think it's a bit nicer to use the target_compile_definitions
```suggestion
  target_compile_definitions(bpftrace_test PRIVATE ARCH_AARCH64)
```

lostjeffle replied:
Actually 'target_compile_definitions' is used in the original patch, but failed with several automatic build test triggered by GitHub, warning that "target_compile_definitions is not defined". Maybe it's related to the version of cmake, since target_compile_definitions is added since cmake3?

fbs replied:
Do you have a link to the failed builds? It should work. We use it in a quite a few places and we require cmake 3.something anyway:

```
tests/CMakeLists.txt
69:  target_compile_definitions(bpftrace_test PRIVATE HAVE_LIBBPF_MAP_BATCH)
73:  target_compile_definitions(bpftrace_test PRIVATE HAVE_NAME_TO_HANDLE_AT=1)
76:  target_compile_definitions(bpftrace_test PRIVATE HAVE_BCC_PROG_LOAD)
79:  target_compile_definitions(bpftrace_test PRIVATE HAVE_BCC_CREATE_MAP)
```

Maybe you placed it before the bpftrace_test target definition? Needs to go on line 69 or later

> I'm not familiar with cmake and really look forward you guys' professional opinion. Thanks.

We all struggle with cmake ;)

lostjeffle commented:
@fbs @danobi 
Changes since v1:
- remove changes in CHANGELOG.md
- use 'target_compile_definitions' instead
- skip some arch specific codegen test cases when arch not supports

fbs commented:
thanks!

If we add a lot of cases it might be nice to change the expected value instead of just removing the test but this is fine for now :)
